src/servers/http_server_v2.cc

```diff
@@ -1101,32 +1101,72 @@ HTTPAPIServerV2::EVBufferToInput(
     const std::string& model_name, const InferRequestHeader& request_header,
     evbuffer* input_buffer,
     TRTSERVER_InferenceRequestProvider* request_provider,
-    AllocPayload* alloc_payload)
+    AllocPayload* alloc_payload, size_t header_length)
 {
   // Extract individual input data from HTTP body and register in
   // 'request_provider'. The input data from HTTP body is not
-  // necessarily contiguous so need to copy into a contiguous
-  // buffer to be able to parse it into a json.
+  // necessarily contiguous so may need to register multiple input
+  // "blocks" for a given input.
   //
-  // Get the addr and size of from the evbuffer.
-  int buffer_len = evbuffer_get_length(input_buffer);
-  char json_buffer[buffer_len];
-  ev_ssize_t extracted_size =
-      evbuffer_copyout(input_buffer, &json_buffer, buffer_len);
-  if (extracted_size != buffer_len) {
+  // Get the addr and size of each chunk of input data from the
+  // evbuffer.
+  struct evbuffer_iovec* v = nullptr;
+  int v_idx = 0;
+
+  int n = evbuffer_peek(input_buffer, -1, NULL, NULL, 0);
+  if (n > 0) {
+    v = static_cast<struct evbuffer_iovec*>(
+        alloca(sizeof(struct evbuffer_iovec) * n));
+    if (evbuffer_peek(input_buffer, -1, NULL, v, n) != n) {
+      return TRTSERVER_ErrorNew(
+          TRTSERVER_ERROR_INTERNAL, "unexpected error getting input buffers");
+    }
+  }
+
+  if (header_length == 0) {
+    return TRTSERVER_ErrorNew(
+        TRTSERVER_ERROR_INTERNAL, "unexpected error header length of size 0");
+  }
+
+  // Extract just the json header from the complete buffer
+  char json_buffer[header_length];
```

Comment:
Use std::vector<char>, not char[]. But you don't need to make a copy at all if the first evbuffer is large enough to contain the entire thing. You likely should move the header extraction logic into a different function. On return you just have to know the updated v_idx (and a v_offset which tells you have many bytes in that block are already consumed from the header).

Reply:
As you specified, I need to null-terminate the string. This means that even if the first evbuffer is large enough to contain the entire thing, if it contains binary data, I will have to create a copy and null terminate that copy at the point when the json content ends.